Stackdriver registry overwrites existing metric descriptors if it fails to pre-populate them, destroying configured descriptions and names

**Describe the bug**
The Stackdriver registry overwrites existing metric descriptors when it fails to pre-populate the known metrics during initialization due to a timeout, causing loss of Terraform-configured data like user-friendly names and descriptions. 

This behavior causes Terraform to attempt recreation of the descriptor via delete and re-create, which can time-out and fail deployments. (related bug ticket https://github.com/hashicorp/terraform-provider-google/issues/22949).

**Environment**
<!-- In what environment did the bug happen? -->
<!-- If you are not using the latest patch version of a supported Micrometer line, please upgrade to see if the issue happens on the latest patch version for that line (e.g. 1.6.x). See https://micrometer.io/support/ -->
 - Micrometer version: 1.14.6
 - Micrometer registry: Stackdriver / Google Cloud Monitoring
 - OS: Linux
 - Java version: 21

**To Reproduce**
Configure a project that reports using Stackdriver
Create e.g. a gauge and set a value.

Eventually when publishing the value, [`createMetricDescriptorIfNecessary`](https://github.com/micrometer-metrics/micrometer/blob/6c7cf24a1415f11dc3ec904fb916d60a316c9f31/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java#L402) will be called. Because this is the first call, `verifiedDescriptors` is empty, and `prePopulateVerifiedDescriptors` [will be called](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java#L426).

`prePopulateVerifiedDescriptors` attempts to fetch existing descriptors via the API client [here](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java#L472). This call needs to fail to trigger the bug (this happens for us sometimes due to a network timeout). The failed call is explicitly ignored by the code [here](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java#L483), and the `verifiedDescriptors` list remains empty.

The calling `createMetricDescriptorIfNecessary` will now not find the metric descriptor in the still-empty `verifiedDescriptors` list, causing a call to create the metric despite it already existing, overwriting it with the incomplete metric information present in the service, which lacks e.g. the friendly name and description.

**Expected behavior**
Existing metric descriptors do not get overwritten.

If the original author's intent to have failures in `prePopulateVerifiedDescriptors` not be showstoppers should be preserved, then I believe a good resolution would be to make the auto-creation of metric descriptors optional via configuration, and if disabled to just attempt to write the time series.

**Additional context**
Log excerpt (slightly anonymized):
```
{
  "insertId": "ID",
  "jsonPayload": {
    "logger": "io.micrometer.stackdriver.StackdriverMeterRegistry",
    "context": "SERVICE NAME",
    "message": "Failed to pre populate verified descriptors for <GCP PROJECT>\ncom.google.api.gax.rpc.DeadlineExceededException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Deadline CallOptions will be exceeded in 29.999807724s. \n\tat com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:94)\n\tat com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)\n\tat com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)\n\tat com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)\n\tat com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)\n\tat com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:84)\n\tat com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1125)\n\tat com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)\n\tat com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1004)\n\tat com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:767)\n\tat com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:516)\n\tat io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:651)\n\tat io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:621)\n\tat io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)\n\tat io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)\n\tat io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)\n\tat com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:569)\n\tat io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)\n\tat io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)\n\tat io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)\n\tat com.google.api.gax.grpc.GrpcLoggingInterceptor$1$1.onClose(GrpcLoggingInterceptor.java:98)\n\tat io.grpc.internal.DelayedClientCall$CloseListenerRunnable.runInContext(DelayedClientCall.java:432)\n\tat io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n\tSuppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed\n\t\tat com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)\n\t\tat com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)\n\t\tat com.google.cloud.monitoring.v3.MetricServiceClient.listMetricDescriptors(MetricServiceClient.java:912)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry$Batch.prePopulateVerifiedDescriptors(StackdriverMeterRegistry.java:473)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry$Batch.createMetricDescriptorIfNecessary(StackdriverMeterRegistry.java:427)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry$Batch.createTimeSeries(StackdriverMeterRegistry.java:402)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry$Batch.createTimeSeries(StackdriverMeterRegistry.java:368)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry.createCounter(StackdriverMeterRegistry.java:276)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry.lambda$publish$2(StackdriverMeterRegistry.java:202)\n\t\tat io.micrometer.core.instrument.Meter.match(Meter.java:104)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry.lambda$publish$10(StackdriverMeterRegistry.java:202)\n\t\tat java.base/java.util.stream.ReferencePipeline$7$1.accept(Unknown Source)\n\t\tat java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)\n\t\tat java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)\n\t\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)\n\t\tat java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)\n\t\tat java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)\n\t\tat java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)\n\t\tat io.micrometer.stackdriver.StackdriverMeterRegistry.publish(StackdriverMeterRegistry.java:207)\n\t\tat io.micrometer.core.instrument.push.PushMeterRegistry.publishSafelyOrSkipIfInProgress(PushMeterRegistry.java:64)\n\t\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)\n\t\tat java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)\n\t\tat java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)\n\t\t... 3 common frames omitted\nCaused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Deadline CallOptions will be exceeded in 29.999807724s. \n\tat io.grpc.Status.asRuntimeException(Status.java:532)\n\t... 14 common frames omitted\n",
    "thread": "stackdriver-metrics-publisher"
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
      "service_name": "SERVICE_NAME",
      "revision_name": "SERVICE_NAME-REVISION",
      "project_id": "PROJECT_ID",
      "location": "europe-west3",
      "configuration_name": "SERVICE_NAME"
    }
  },
  "timestamp": "2025-05-21T03:54:20.991929322Z",
  "severity": "WARNING",
  "labels": {
    "instanceId": "ID"
  },
  "logName": "projects/PROJECT_ID/logs/run.googleapis.com%2Fstdout",
  "receiveTimestamp": "2025-05-21T03:54:21.211065676Z"
}
```
[...] Other errors ommitted
```
{
  "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "authenticationInfo": {
      "principalEmail": "SERVICE_NAME-runner@PROJECT_ID.iam.gserviceaccount.com",
      "serviceAccountDelegationInfo": [
        {
          "firstPartyPrincipal": {
            "principalEmail": "service-XXXXXXXXXXX@serverless-robot-prod.iam.gserviceaccount.com"
          }
        }
      ],
      "principalSubject": "serviceAccount:SERVICE-runner@PROJECT_ID.iam.gserviceaccount.com"
    },
    "requestMetadata": {
      "callerIp": "10.5.3.10",
      "callerSuppliedUserAgent": "Spring/6.2.0 spring-cloud-gcp-metrics/6.2.0 grpc-java-netty/1.70.0,gzip(gfe)",
      "callerNetwork": "//compute.googleapis.com/projects/PROJECT_ID/global/networks/__unknown__",
      "requestAttributes": {
        "time": "2025-05-21T03:54:27.605392688Z",
        "auth": {}
      },
      "destinationAttributes": {}
    },
    "serviceName": "monitoring.googleapis.com",
    "methodName": "google.monitoring.v3.MetricService.CreateMetricDescriptor",
    "authorizationInfo": [
      {
        "resource": "projects/PROJECT_ID",
        "permission": "monitoring.metricDescriptors.create",
        "granted": true,
        "resourceAttributes": {},
        "permissionType": "ADMIN_WRITE"
      }
    ],
    "resourceName": "projects/PROJECT_ID",
    "request": {
      "name": "projects/PROJECT_ID",
      "@type": "type.googleapis.com/google.monitoring.v3.CreateMetricDescriptorRequest",
      "metricDescriptor": {
        "type": "custom.googleapis.com/METRIC_NAME",
        "valueType": "DOUBLE",
        "metricKind": "CUMULATIVE"
      }
    },
    "response": {
      "valueType": "DOUBLE",
      "name": "projects/PROJECT_ID/metricDescriptors/custom.googleapis.com/METRIC_NAME",
      "@type": "type.googleapis.com/google.api.MetricDescriptor",
      "type": "custom.googleapis.com/METRIC_NAME",
      "metricKind": "CUMULATIVE"
    }
  },
  "insertId": "ID",
  "resource": {
    "type": "audited_resource",
    "labels": {
      "service": "monitoring.googleapis.com",
      "project_id": "PROJECT_ID",
      "method": "google.monitoring.v3.MetricService.CreateMetricDescriptor"
    }
  },
  "timestamp": "2025-05-21T03:54:27.597394162Z",
  "severity": "NOTICE",
  "logName": "projects/PROJECT_ID/logs/cloudaudit.googleapis.com%2Factivity",
  "receiveTimestamp": "2025-05-21T03:54:29.185216918Z"
}
```

--- a/implementations/micrometer-registry-stackdriver/build.gradle
+++ b/implementations/micrometer-registry-stackdriver/build.gradle
@@ -11,6 +11,7 @@ dependencies {
     implementation libs.slf4jApi
     compileOnly libs.logback12
     testRuntimeOnly libs.logback12
+    testImplementation libs.mockitoCore5
     // needed for extending TimeWindowPercentileHistogram in StackdriverHistogramUtil
     compileOnly libs.hdrhistogram
 

--- a/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/MetricServiceClientFactory.java
+++ b/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/MetricServiceClientFactory.java
@@ -1,0 +1,27 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.stackdriver;
+
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.monitoring.v3.MetricServiceSettings;
+
+import java.io.IOException;
+
+interface MetricServiceClientFactory {
+
+    MetricServiceClient create(MetricServiceSettings settings) throws IOException;
+
+}

--- a/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverConfig.java
+++ b/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverConfig.java
@@ -121,6 +121,16 @@ public interface StackdriverConfig extends StepRegistryConfig {
         }).get();
     }
 
+    /**
+     * Whether it should be attempted to create a metric descriptor before writing a time
+     * series. If you set up your metric descriptors using IaC tools like Terraform, you
+     * will likely want to disable this.
+     * @return true, if metric descriptors should be auto-created
+     */
+    default boolean autoCreateMetricDescriptors() {
+        return getBoolean(this, "autoCreateMetricDescriptors").orElse(true);
+    }
+
     @Override
     default Validated<?> validate() {
         return checkAll(this, c -> StepRegistryConfig.validate(c),

--- a/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java
+++ b/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMeterRegistry.java
@@ -73,6 +73,8 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
 
     private final StackdriverConfig config;
 
+    private final MetricServiceClientFactory clientFactory;
+
     private long previousBatchEndTime;
 
     /**
@@ -87,14 +89,16 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
     MetricServiceClient client;
 
     public StackdriverMeterRegistry(StackdriverConfig config, Clock clock) {
-        this(config, clock, DEFAULT_THREAD_FACTORY, () -> MetricServiceSettings.newBuilder().build());
+        this(config, clock, DEFAULT_THREAD_FACTORY, () -> MetricServiceSettings.newBuilder().build(),
+                new StackdriverMetricServiceClientFactory());
     }
 
     private StackdriverMeterRegistry(StackdriverConfig config, Clock clock, ThreadFactory threadFactory,
-            Callable<MetricServiceSettings> metricServiceSettings) {
+            Callable<MetricServiceSettings> metricServiceSettings, MetricServiceClientFactory clientFactory) {
         super(config, clock);
 
         this.config = config;
+        this.clientFactory = clientFactory;
 
         try {
             this.metricServiceSettings = metricServiceSettings.call();
@@ -123,7 +127,7 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
             else {
                 shutdownClientIfNecessary(true);
                 try {
-                    this.client = MetricServiceClient.create(metricServiceSettings);
+                    this.client = this.clientFactory.create(metricServiceSettings);
                     super.start(threadFactory);
                 }
                 catch (Exception e) {
@@ -311,6 +315,8 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
 
         private Callable<MetricServiceSettings> metricServiceSettings;
 
+        private MetricServiceClientFactory clientFactory = new StackdriverMetricServiceClientFactory();
+
         Builder(StackdriverConfig config) {
             this.config = config;
             this.metricServiceSettings = () -> {
@@ -338,8 +344,14 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
             return this;
         }
 
+        // VisibleForTesting
+        Builder clientFactory(MetricServiceClientFactory clientFactory) {
+            this.clientFactory = clientFactory;
+            return this;
+        }
+
         public StackdriverMeterRegistry build() {
-            return new StackdriverMeterRegistry(config, clock, threadFactory, metricServiceSettings);
+            return new StackdriverMeterRegistry(config, clock, threadFactory, metricServiceSettings, clientFactory);
         }
 
     }
@@ -421,6 +433,10 @@ public class StackdriverMeterRegistry extends StepMeterRegistry {
         private void createMetricDescriptorIfNecessary(MetricServiceClient client, Meter.Id id,
                 MetricDescriptor.ValueType valueType, @Nullable String statistic,
                 MetricDescriptor.MetricKind metricKind) {
+
+            if (!config.autoCreateMetricDescriptors()) {
+                return;
+            }
 
             if (verifiedDescriptors.isEmpty()) {
                 prePopulateVerifiedDescriptors();

--- a/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMetricServiceClientFactory.java
+++ b/implementations/micrometer-registry-stackdriver/src/main/java/io/micrometer/stackdriver/StackdriverMetricServiceClientFactory.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.stackdriver;
+
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.monitoring.v3.MetricServiceSettings;
+
+import java.io.IOException;
+
+class StackdriverMetricServiceClientFactory implements MetricServiceClientFactory {
+
+    @Override
+    public MetricServiceClient create(MetricServiceSettings settings) throws IOException {
+        return MetricServiceClient.create(settings);
+    }
+
+}

--- a/implementations/micrometer-registry-stackdriver/src/test/java/io/micrometer/stackdriver/StackdriverMeterRegistryDescriptorAutoCreationTest.java
+++ b/implementations/micrometer-registry-stackdriver/src/test/java/io/micrometer/stackdriver/StackdriverMeterRegistryDescriptorAutoCreationTest.java
@@ -1,0 +1,106 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.stackdriver;
+
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.monitoring.v3.MetricServiceSettings;
+import com.google.monitoring.v3.ListMetricDescriptorsRequest;
+import io.micrometer.core.instrument.Tags;
+import org.jspecify.annotations.Nullable;
+import org.junit.jupiter.api.Test;
+
+import java.util.concurrent.Executors;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+class StackdriverMeterRegistryDescriptorAutoCreationTest {
+
+    @Test
+    void disabledAutoCreationShouldNotCreateMetricDescriptor() {
+        var mockClient = mock(MetricServiceClient.class);
+        when(mockClient.listMetricDescriptors((ListMetricDescriptorsRequest) any())).thenReturn(emptyResponse());
+
+        var meterRegistry = mockMeterRegistry(mockClient, false);
+
+        meterRegistry.start(Executors.defaultThreadFactory());
+
+        meterRegistry.counter("testCounter", Tags.empty());
+
+        meterRegistry.publish();
+
+        verify(mockClient, never()).createMetricDescriptor(any());
+        verify(mockClient, times(1)).createTimeSeries(any());
+    }
+
+    @Test
+    void enabledAutoCreationShouldCreateMetricDescriptor() {
+        var mockClient = mock(MetricServiceClient.class);
+        var meterRegistry = mockMeterRegistry(mockClient, true);
+
+        meterRegistry.start(Executors.defaultThreadFactory());
+
+        meterRegistry.counter("testCounter", Tags.empty());
+
+        meterRegistry.publish();
+
+        verify(mockClient, times(1)).createMetricDescriptor(any());
+        verify(mockClient, times(1)).createTimeSeries(any());
+    }
+
+    private StackdriverMeterRegistry mockMeterRegistry(MetricServiceClient mockClient,
+            boolean autoCreateMetricDescriptors) {
+        var factory = new MetricServiceClientFactory() {
+            @Override
+            public MetricServiceClient create(MetricServiceSettings settings) {
+                return mockClient;
+            }
+        };
+
+        return StackdriverMeterRegistry.builder(stackdriverConfig(autoCreateMetricDescriptors))
+            .clientFactory(factory)
+            .build();
+    }
+
+    private StackdriverConfig stackdriverConfig(boolean autoCreateMetricDescriptors) {
+        return new StackdriverConfig() {
+            @Override
+            public boolean enabled() {
+                return true;
+            }
+
+            @Override
+            public String projectId() {
+                return "doesnotmatter";
+            }
+
+            @Override
+            public @Nullable String get(String key) {
+                return null;
+            }
+
+            @Override
+            public boolean autoCreateMetricDescriptors() {
+                return autoCreateMetricDescriptors;
+            }
+        };
+    }
+
+    private MetricServiceClient.ListMetricDescriptorsPagedResponse emptyResponse() {
+        return mock(MetricServiceClient.ListMetricDescriptorsPagedResponse.class);
+    }
+
+}

